android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/Keyboard.java

```diff
@@ -21,6 +21,38 @@
   private String font;
   private String oskFont;
 
+  // JSON keys
+  public static String KB_NEW_KEYBOARD_KEY = "isNewKeyboard";
+  public static String KB_FONT_KEY = "font";
+  public static String KB_OSK_FONT_KEY = "oskFont";
+
+  /**
+   * Constructor using JSON Objects from installed keyboards list
+   * @param installedObj
+   */
+  public Keyboard(JSONObject installedObj) {
+    try {
+      this.packageID = installedObj.getString(LanguageResource.LR_PACKAGE_ID_KEY);
+      this.resourceID = installedObj.getString(LanguageResource.LR_RESOURCE_ID_KEY);
+      this.resourceName = installedObj.getString(LanguageResource.LR_RESOURCE_NAME_KEY);
+      this.languageID = installedObj.getString(LanguageResource.LR_LANGUAGE_ID_KEY);
+      this.languageName = installedObj.getString(LanguageResource.LR_LANGUAGE_NAME_KEY);
+      this.version = installedObj.getString(LanguageResource.LR_VERSION_KEY);
+      this.helpLink = installedObj.getString(LanguageResource.LR_HELP_LINK_KEY);
```

Comment:
This block appears to belong better in the parent class. Is it possible to refactor accordingly so that the concerns are consistently separated?

It would also be good to introduce a protected `fromJSON()` and put the JSON loading code in that, calling it from this constructor, to mirror the `toJSON()` method you have created here. Something like:

```
public Keyboard(JSONObject installedObj) {
   this.fromJSON(installedObj);
}

protected fromJSON(JSONObject installedObj) {
  super.fromJSON(installedObj);
  try {
    this.isNewKeyboard = installedObj.getBoolean(KB_NEW_KEYBOARD_KEY);
    ...
  } catch(JSONException e) ...
```

Reply:
Constructors refactored. Unfortunately, I had to add stubs for two other constructors.